loggly/loggly.go

```diff
@@ -76,59 +79,80 @@ func (l *Adapter) Stream(logstream chan *router.Message) {
 			ContainerImage:    m.Container.Config.Image,
 			ContainerHostname: m.Container.Config.Hostname,
 		}
-
-		l.SendMessage(msg)
+		a.enquequeMessage(msg)
 	}
 }
 
-// SendMessage handles creating and sending a request to Loggly. Any errors
-// that occur during that process are bubbled up to the caller
-func (l *Adapter) SendMessage(msg logglyMessage) {
-	js, err := json.Marshal(msg)
+// dispatcher sends bulks of bufferd logs every 2 seconds after a request to loggly
+func (a *Adapter) dispatcher() {
+	for {
+		time.Sleep(2 * time.Second)
+		if len(a.msgs) == 0 {
+			continue
+		}
+		a.dequeueMessages()
+	}
+}
+func (a *Adapter) enquequeMessage(msg logglyMessage) {
+	a.m.Lock()
+	a.msgs = append(a.msgs, msg)
+	a.m.Unlock()
+}
 
+// dequeueMessages marshals the log messages into json and cleas the buffer
+func (a *Adapter) dequeueMessages() {
+	a.m.Lock()
+	msgs := a.msgs
+	// clear log buffer
+	a.msgs = []logglyMessage{}
+	a.m.Unlock()
+
+	buf := &bytes.Buffer{}
+	for i, msg := range msgs {
+		msgJson, err := json.Marshal(msg)
+		if err != nil {
+			a.log.Println(err)
+			continue
+		}
+		_, err = buf.Write(msgJson)
+		if err != nil {
+			a.log.Println(err)
+			continue
+		}
+		// split the messages with a new line exept the last one
+		if i < len(msgs)-1 {
+			buf.WriteRune('\n')
+		}
+	}
+	err := a.sendLogs(buf)
 	if err != nil {
-		log.Println(err)
-		return
+		// if the send fialed requeue the log messages and send next time
+		log.Println("failed to rend logs to loggly, requeueing: ", err)
```

Comment:
So if this image is not configured correctly or Loggly is down the memory use of this image will increase for ever? We should try to make a best effort to get messages to loggly but I don't think we should always requeue messages. I think  a retry of N times (configurable) would be appropriate